### ISSUE TITLE
Merge string_length and array_length into length() and make it work for map-like objects

### ISSUE THREAD

**Issue by rowanseymour** [nyaruka/goflow#217]
Currently you can do:

```
@(array_length(contact.urns), 1)
@(array_length(split("a, b, c")), 3)
@(string_length(split("a, b, c")), 7) // not so helpful...
```

Seems like it would be nice to be also able to get the length of some things which aren't indexable, e.g. run results. We could split the `Indexable` interface into...

```golang
type Lengthable interface { // open to name suggestions! 
  Length() int
}

type Indexable interface {
  Lengthable 
  Index(int) interface{}
}
```

and implement it on `run.results`, `contact.fields`, `JSONFragment` etc. Rather than introduce an additional length method to do these I think we should have a single `length()` which works with strings, indexables (including arrays) and other lengthables (like `run.results`)

```
@(length("a, b, c"), 7)         // string
@(length(split("a, b, c")), 3)  // array
@(length(contact.urns), 1)      // indexable complex object
@(length(run.results), 5)       // lengthable but not indexable complex object
```

I get the argument for forcing users to be explicit about types but I think that makes more sense in cases like `+` vs `&` where the operation is different depending on the types. There's definitely a place for methods like `to_json` which can take different types but they return the same thing.

**Comment by nicpottier** [comment]
I think there's two arguments for the type specific methods:
 1. The user's intent is unambiguous, so if we ever want to rewrite things or rename things in the future, we can do that easily without introducing `legacy_length`.
 2. Though it is clear to US what type something is, it isn't super transparent to the user (an array of elements can look exactly like string of elements in every way they can see), so type asserting methods make sure they are getting what they want. (and communicate that to us so we can tell them if they are wrong about their thoughts on what that type is)

Length feels pretty safe from a 1) point of view, I think 2) is potentially a concern. I think 2) could be especially troublesome if we ever decide to do fancy things with `String()` on items that are Lengthable. IE, if we had something like `contact.urns` resolve to the top level URN, then in the context the user things `contact.urns` resolves to `tel:+12067799291`, but when they call length on `contact.urns` then then get 1 instead of 14 or whatever, and that happens silently.

As I recall we decided to introduce `contact.urn` so that is no longer a concern, but I think moving to a universal length means we need to make sure all Indexable items always have their String resolve to a "listy"/"mappy" looking thing instead of doing anything fancy as far as showing the highest / latest / top of a list/map.

**Comment by rowanseymour** [comment]
I hear ya on concern 2) and I think moving away from "clever" `Default()` values is a sensible improvement in that regard.

Here's another length-related thing to ponder - what is `length(1.0)`. We've always stringified the input and took the length of that... but should we? We need to be generous about interpreting strings as more type-specific values - after all results are always strings, and contact fields are often strings containing numbers and dates. But should we be so generous in the other direction? Maybe `length()` should only accept a string or a `Lengthable` *

* In the future string might be `XString` which implements `Lengthable`?

**Comment by rowanseymour** [comment]
Addressed in #223
